### ISSUE TITLE
Package "signed by a HashiCorp partner" yet "registry.terraform.io does not have a provider named registry.terraform.io/hashicorp/gitlab"

### ISSUE THREAD

**Issue by jgspratt** [hashicorp/terraform#28298]
[The Registry page for the Gitlab provider](https://registry.terraform.io/providers/gitlabhq/gitlab/latest) says to use the following configuration to get the provider:

```
terraform {
  required_providers {
    gitlab = {
      source = "gitlabhq/gitlab"
      version = "3.6.0"
    }
  }
}
```

I therefore included it as follows:

```
terraform {
  required_providers {
    null = {
      source  = "registry.terraform.io/hashicorp/null"
      version = "~> 3.1"
    }
    gitlab = {
      source  = "gitlabhq/gitlab"
      version = "3.6.0"
    }
  }
}
```

When mirroring this to the local plugins directory, I get the following failure:

```
# terraform providers mirror ~/.terraform.d/plugins

- Mirroring gitlabhq/gitlab...
  - Selected v3.6.0 to meet constraints 3.6.0
  - Downloading package for linux_amd64...
  - Package authenticated: signed by a HashiCorp partner
- Mirroring hashicorp/gitlab...
- Mirroring hashicorp/null...
  - Selected v3.1.0 to meet constraints ~> 3.1
  - Downloading package for linux_amd64...
  - Package authenticated: signed by HashiCorp

Error: Provider not available

Failed to download registry.terraform.io/hashicorp/gitlab from its origin
registry: provider registry registry.terraform.io does not have a provider
named registry.terraform.io/hashicorp/gitlab.
```

How can it see that the package is `signed by a HashiCorp partner` but that the package doesn't exist?

I have also tried `source  = "registry.terraform.io/gitlabhq/gitlab"` but it results in the same error.

Something appears to get downloaded:

```
# md5sum ~/.terraform.d/plugins/registry.terraform.io/gitlabhq/gitlab/terraform-provider-gitlab_3.6.0_linux_amd64.zip
07e4bd3b78fbf241b27204b14756859b  /home/gitlab-runner/.terraform.d/plugins/registry.terraform.io/gitlabhq/gitlab/terraform-provider-gitlab_3.6.0_linux_amd64.zip

# sha256sum ~/.terraform.d/plugins/registry.terraform.io/gitlabhq/gitlab/terraform-provider-gitlab_3.6.0_linux_amd64.zip
5a795890e78121f7b8d8a3ae6fdc6a6f69083465b6a2073c9c0f0564dad4f7ec  /home/gitlab-runner/.terraform.d/plugins/registry.terraform.io/gitlabhq/gitlab/terraform-provider-gitlab_3.6.0_linux_amd64.zip

# cat ~/.terraform.d/plugins/registry.terraform.io/gitlabhq/gitlab/3.6.0.json
{
  "archives": {
    "linux_amd64": {
      "hashes": [
        "h1:S0PE1Jk6gNtGetAJSj8ETPyRBA9a++647YAPeP47H0s="
      ],
      "url": "terraform-provider-gitlab_3.6.0_linux_amd64.zip"
    }
  }
}

# cat ~/.terraform.d/plugins/registry.terraform.io/gitlabhq/gitlab/index.json
{
  "versions": {
    "3.6.0": {}
  }
}
```

### Terraform Version

```
Terraform v0.14.9
# uname -a
Linux gitlab.internal 4.18.0-294.el8.x86_64 #1 SMP Mon Mar 15 22:38:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jgspratt** [comment]
Ah, got it.

I wasn't setting the provider on the module I was using, so it was defaulting to the hashicorp namespace.  I'm not used to modules needing provider requirements.

This clued me in:

https://discuss.hashicorp.com/t/terraform-init-failed-to-query-available-provider-packages-looking-for-registry-terraform-io-hashicorp-fastly/19486/9

**Comment by alisdair** [comment]
Glad you figured it out! This is a common stumbling block, and we've improved the messaging on this recently in #28014, which will be released as part of 0.15.0.
